### PR TITLE
GitHub Actions CI

### LINKED ISSUE #15: GitHub Actions
# GitHub Actions

## Summary

We should switch from travis to GitHub Actions.

## Motivation

More native support within GitHub.

## Detailed Design

At first we should just migrate our travis file to GitHub Actions.  
We should also update used node versions. (Target LTS = 12 and latest = 14).
Maybe we can also add a badge that represent the latest build status.

## Documentation

Maybe? In a wiki?

## Drawbacks

Dont think so.

## Alternatives

We could stay at travis, but why?

## Open Questions

We maybe need to think about coverage steps and uploading artefacts.

## Help Needed

I have already collected some experience with GitHub Actions 🙂

## Frequently Asked Questions

Currently none

## Related Discussions

Currently none

### PR DESCRIPTION
**Pull request by Shinigami92**
closes #15

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      #   run: yarn lint
+
+      - name: Audit dependencies
+        run: yarn audit --groups dependencies
+
+      # - name: Audit devDependencies
+      #   run: yarn audit --groups devDependencies
+
+      # - name: Audit peerDependencies
+      #   run: yarn audit --groups peerDependencies
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:42]: `peerDependencies` are currently empty

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      # - name: Audit optionalDependencies
+      #   run: yarn audit --groups optionalDependencies
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:45]: `optionalDependencies` are currently empty

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Build
+        run: yarn build
+
+      - name: Test
+        run: yarn test
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:51]: `test` are currently failing, cause the project is 2 years old (I think 😅)

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Codecov
+        run: yarn codecov
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:54]: Codecov is requesting a repo token

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+      - name: Codeclimate
+        run: yarn codeclimate-test-reporter < ./coverage/lcov.info
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:57]: Codeclimate requests `CODECLIMATE_REPO_TOKEN`

Comment (Shinigami92) [inline .github/workflows/ci.yml:57]: Don't know if `< ./coverage/lcov.info` is working in GitHub Actions the same way as in travis

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Set node version to ${{ matrix.node }}
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node }}
+
+      - name: Versions
+        run: yarn versions
+
+      - name: Install dependencies
+        run: yarn install #--frozen-lockfile
```
Comment (Shinigami92) [inline .github/workflows/ci.yml:27]: We can enable `--frozen-lockfile` if #16 was solved

File: .travis.yml [hunk 7]
```diff
@@ -1,20 +0,0 @@
...
-  directories:
-    - node_modules
-language: node_js
-node_js:
-  - node
-  - 8
-notifications:
-  webhooks:
-    urls:
-      - "https://webhooks.gitter.im/e/94a8a87b117fbc50ceaf"
```
Comment (Shinigami92) [inline .travis.yml:14]: I think this can be supported via `Settings > Webhooks`

Comment (zimme) [reply .travis.yml:14]: This is super old the Gitter rooms are removed so you can disregard this.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (zimme) [reply .github/workflows/ci.yml:27]: Let's not enable this until we settle on what to do with #16.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (zimme) [reply .github/workflows/ci.yml:57]: I don't see why not as that's standard shell syntax.

Comment (zimme) [reply .github/workflows/ci.yml:57]: I'll give you the token in private and let's check if you have access to code climate as they use GitHub backed auth.

Comment (Shinigami92) [reply .github/workflows/ci.yml:57]: mh 🤔 https://docs.codeclimate.com/docs/command-line-interface#error-no-codeclimate_repo_token-found-a-codeclimate_repo_token-must-be-specified-as-an-environment-variable

Do we already need it? Or is the dependency outdated?

File: .github/workflows/ci.yml (hunk 4 above)
Comment (zimme) [reply .github/workflows/ci.yml:54]: Just checked travis and I don't seem to have had the codecov token setup, maybe I just ran codecov to get a local coverage folder that I could use with code climate?

Comment (Shinigami92) [reply .github/workflows/ci.yml:54]: Yes, maybe

We could also think about a general new code coverage solution at all

File: .github/workflows/ci.yml (hunk 3 above)
Comment (zimme) [reply .github/workflows/ci.yml:51]: I believe I was working on fixing some tests when I stopped working on this project and the problem was something related to not being able to mock node's module resolution on the lower level I needed to mock it... but this is more about moving from travis to GH actions so I'm ok with ignoring failing tests for now.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (Shinigami92) [reply .github/workflows/ci.yml:57]: I would suggest to update and bring the project to the newest state in a dedicated PR

There I will update all package deps and try to fix all error's to bring the current state of the project to a working state

_This could also help me onboarding into what the current idea of the project is/was_

Comment (zimme) [reply .github/workflows/ci.yml:57]: Sounds like a plan.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (zimme) [reply .github/workflows/ci.yml:54]: Let's fix the codecov/codeclimate thing in a separate PR

**Comment by zimme** [comment]
Let's remove the codecov/codeclimate parts of the CI pipeline for now and add them again in a separate PR so we can set them up properly using the current best practices.

**Pull request merged**